### ISSUE TITLE
Indexing into tensor order of magnitude slower than numpy

### ISSUE THREAD

**Issue by DomHudson** [pytorch/pytorch#29973]
## 🐛 Bug

Indexing into a pytorch tensor is an order of magnitude slower than numpy.

## To Reproduce

```python
import torch
import numpy as np


BATCH_SIZE = 32
SEQUENCE_LENGTH = 512

TORCH_MATRIX = torch.full(
    size = (BATCH_SIZE, SEQUENCE_LENGTH),
    fill_value = 0,
    dtype = int,
)

NUMPY_MATRIX = np.full(
    shape = (BATCH_SIZE, SEQUENCE_LENGTH),
    fill_value = 0,
    dtype = int,
)
    

def index_over_matrix(matrix):
    for row_index in range(BATCH_SIZE):
        for column_index in range(SEQUENCE_LENGTH):
            matrix[row_index][column_index]
```

```
%%timeit -n 30

index_over_matrix(NUMPY_MATRIX)
```
>>> 5.9 ms ± 917 µs per loop (mean ± std. dev. of 7 runs, 30 loops each)

```
%%timeit -n 30

index_over_matrix(TORCH_MATRIX)
```
>>> 50.3 ms ± 2.5 ms per loop (mean ± std. dev. of 7 runs, 30 loops each)

## Line profiler

```
Timer unit: 1e-06 s

Total time: 0.143154 s
File: <ipython-input-16-108e7beab457>
Function: index_over_matrix at line 21

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    21                                           def index_over_matrix(matrix):
    22        33         48.0      1.5      0.0      for row_index in range(BATCH_SIZE):
    23     16416      14083.0      0.9      9.8          for column_index in range(SEQUENCE_LENGTH):
    24     16384     129023.0      7.9     90.1              matrix[row_index][column_index]
``` 

## Expected behavior

I expect indexing to be quite quick.

## Environment

```
PyTorch version: 1.3.0
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] msgpack-numpy==0.4.4.3
[pip3] numpy==1.17.4
[pip3] pytorch-lamb==1.0.0
[pip3] torch==1.3.0
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.4.1
[pip3] torchviz==0.0.1
[conda] Could not collect
```

cc @ezyang @SsnL @albanD @zou3519 @gqchen @VitalyFedyunin @ngimel @gchanan @jerryzh168 @mruberry

**Comment by rmcavoy** [comment]
I recommend testing the indexing as [i, j] for a comparison. IIRC, [i][j] should slice it into columns and then index the columns so has the potential to be quite a bit heavier.

**Comment by DomHudson** [comment]
Thank you - that is a good tip! Numpy appears faster this way but, the difference with torch still remains.

```python
def index_with_commas_over_matrix(matrix):
    for row_index in range(BATCH_SIZE):
        for column_index in range(SEQUENCE_LENGTH):
            matrix[row_index, column_index]
```

```python
%%timeit -n 30

index_with_commas_over_matrix(NUMPY_MATRIX)
```
>>> 3.56 ms ± 875 µs per loop (mean ± std. dev. of 7 runs, 30 loops each)

```python
%%timeit -n 30

index_with_commas_over_matrix(TORCH_MATRIX)
```
>>> 51.2 ms ± 3.15 ms per loop (mean ± std. dev. of 7 runs, 30 loops each)

**Comment by rmcavoy** [comment]
Does it maintain that speed diff with the TORCH_matrix computed in an environment where autograd should be disabled? i.e. test with
```
with torch.no_grad():
      index_with_commas_over_matrix(TORCH_MATRIX)
```

(I would confirm your test myself but I don't generally run inside Jupyter and unless I am not mistaken that timeit syntax is invalid outside ipython)

**Comment by DomHudson** [comment]
Unfortunately the difference remains on my machine with the `no_grad` context manager - I used 100 loops this time.

```python
%%timeit -n 100

index_with_commas_over_matrix(NUMPY_MATRIX)
```
>>> 3.37 ms ± 353 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```python
%%timeit -n 100

index_with_commas_over_matrix(TORCH_MATRIX)
```
>>> 49.3 ms ± 1.43 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

``` python
%%timeit -n 100

with torch.no_grad():
    index_with_commas_over_matrix(TORCH_MATRIX)
```
>>> 53.2 ms ± 4.26 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

**Comment by mruberry** [comment]
Lowering priority (for the moment). Thanks for the filing this issue! Iterating over a matrix like this seems like an unusual pattern. While PyTorch's indexing does appear to be slower than numpy's, are there some real-world workloads whose performance you think would be significantly impacted by this issue? Or is there a natural coding pattern that can't workaround this issue?

**Comment by fmassa** [comment]
This is related (if not a duplicate) of https://github.com/pytorch/pytorch/issues/5388

**Comment by DomHudson** [comment]
Hi @mruberry !
Thanks for taking a look at this issue. I agree that iterating over all elements like this would be unusual, however I don't think indexing into the tensor is unusual.  For example, it will be common to create a tensor and manipulate its values to create your model input. You could convert a numpy array or python list to the pytorch tensor but it would be reasonable to assume it would be quicker to begin with the correct type (especially as array operations are supported) rather than going through the extra logic to convert the type.
The readme of pytorch claims that the tensors are fast - I think retrieving or altering the value at a given position should be within a reasonable range of other available tools.

Many thanks,
Dom

**Comment by mruberry** [comment]
> Hi @mruberry !
> Thanks for taking a look at this issue. I agree that iterating over all elements like this would be unusual, however I don't think indexing into the tensor is unusual. For example, it will be common to create a tensor and manipulate its values to create your model input. You could convert a numpy array or python list to the pytorch tensor but it would be reasonable to assume it would be quicker to begin with the correct type (especially as array operations are supported) rather than going through the extra logic to convert the type.
> The readme of pytorch claims that the tensors are fast - I think retrieving or altering the value at a given position should be within a reasonable range of other available tools.
> 
> Many thanks,
> Dom

Yep, that makes sense.

**Comment by szan1435** [comment]
> Lowering priority (for the moment). Thanks for the filing this issue! Iterating over a matrix like this seems like an unusual pattern. While PyTorch's indexing does appear to be slower than numpy's, are there some real-world workloads whose performance you think would be significantly impacted by this issue? Or is there a natural coding pattern that can't workaround this issue?

Hi @mruberry ,

The use case of indexing tensor is not unusual. In my case, I would like to slice the weight tensor of a neural network layer based on the L1 norm for pruning. However, the speed for such slicing is very slow. Please refer to the following post on Pytorch forum:

https://discuss.pytorch.org/t/how-to-accelerate-the-process-of-tensor-index-select-and-index-copy/41308/4

https://discuss.pytorch.org/t/efficient-batch-index-select/28305

It would be nice if this issue can be solved. 

Best regards,

Shizhe

**Comment by szan1435** [comment]
> Hi @mruberry !
> Thanks for taking a look at this issue. I agree that iterating over all elements like this would be unusual, however I don't think indexing into the tensor is unusual. For example, it will be common to create a tensor and manipulate its values to create your model input. You could convert a numpy array or python list to the pytorch tensor but it would be reasonable to assume it would be quicker to begin with the correct type (especially as array operations are supported) rather than going through the extra logic to convert the type.
> The readme of pytorch claims that the tensors are fast - I think retrieving or altering the value at a given position should be within a reasonable range of other available tools.
> 
> Many thanks,
> Dom

Hi @DomHudson ,

Thank you very much for pointing this out! I have met the same issue. May I know if you come up with faster approach to index_select a tensor? 

Best regards,

Shizhe

**Comment by DomHudson** [comment]
> Hi @DomHudson ,
> 
> Thank you very much for pointing this out! I have met the same issue. May I know if you come up with faster approach to index_select a tensor?
> 
> Best regards,
> 
> Shizhe

Hi @szan1435,

I ended up running these operations in numpy and then moving it back to pytorch:

```python
return torch.from_numpy(matrix).to(device)
```

This sped up my application significantly. Note that: in my use case, I was constructing tensors from non-matrix input so I didn't actually begin with a pytorch tensor but I needed to end up with one.

If you already have a pytorch tensor and want to index into it, you may find the overhead of moving it to numpy and then back counteracts the indexing speed improvement  - I suppose it depends on how much indexing you need to do! 

I would be interested in your findings.

Best,
Dom

**Comment by szan1435** [comment]
> @szan1435, can you please share a script that you used to produce your profiling results? I'm interested in sizes of your tensors and approximate number of nonzeros in the mask.

Hi @ngimel , I am sorry that the profiling result is wrong. There is no much difference between the tensor indexing and tensor indexing after correction.

**Comment by DomHudson** [comment]
@VitalyFedyunin  Hi, why did you close this? Believe this is still ongoing and used to track https://github.com/pytorch/pytorch/issues/43153 and https://github.com/pytorch/pytorch/issues/43116

**Comment by vwxyzjn** [comment]
@ljk53  that was a very move profiling. Any updates with this?

**Comment by yeshwanthv5** [comment]
Please share if anyone has a temporary hack for this problem.

**Comment by DomHudson** [comment]
@yeshwanthv5  I would suggest moving to numpy and back if you need to do a lot of indexing.

---

I suspect that @szan1435 was mistaken as the author of this ticket and the ticket was closed after his comment that his profiling was wrong. He was only talking about his profiling - mine still illustrates that Numpy is just over 15 times faster than Pytorch per index lookup.

@mruberry @VitalyFedyunin is it possible to reopen this ticket please? If it's closed it may be harder to find it which is a shame as it's currently being used to track several other tickets.

**Comment by akamaus** [comment]
Something is definitely strange. Not only torch's indexing is around ten times slower than numpy's, but time also grows linearly with tensor dimension. In contrast, for numpy access time is roughly stays constant.

```
from time import time
def bench_dim(d):
    R = 100000
    tti = torch.zeros([2]*d, dtype=int)
    nai = np.zeros([2]*d, dtype=int)
    idx = tuple([1] * d)
    
    t0=time()
    for k in range(R):
        pass
    dt_loop = time() - t0
    
    t0=time()
    for k in range(R):
        nai[idx]
    dt_np = time() - t0
    
    t0=time()
    for k in range(R):
        tti[idx]
    dt_torch = time() - t0
    
    return dt_loop / R , dt_np / R, dt_torch / R

ds = []
ts = []
for d in range(1,10):
    t = bench_dim(d)
    ds.append(d)
    ts.append(t)
```

![image](https://user-images.githubusercontent.com/58955/103168464-8322bd00-4844-11eb-8912-224bad7d9ff7.png)

**Comment by akamaus** [comment]
To isolate Python's influence, I replicated measurements in C++ with the following code:
```
#include <torch/torch.h>
#include <iostream>

#include <chrono>

#define R 100000
int main() {
  std::chrono::high_resolution_clock timer;  

  torch::Tensor tensor = torch::zeros({2}).to(torch::kInt);

  auto t0 = timer.now();
  for (int i=0; i<R; i++) {
    tensor[1];
  }
  auto t1 = timer.now();
  std::chrono::duration<double> elapsed = t1-t0;
  std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(elapsed).count() / R << std::endl;
}
```

On the same machine I'm getting 450ns for 1-d tensor and 550ns per access for 10-d tensor. So both raw timings and scaling are considerably better than that for Python. 

Still, replacing a tensor with a raw C array gives access times around 2ns. So using pytorch tensors instead of native arrays is ~20x slower.

**Comment by kumar-akshay324** [comment]
Is anybody working on this or has this been already resolved? I can confirm for larger 1-D tensors of length > 5000, a numpy look up is a magnitude faster than the torch access.

**Comment by pearu** [comment]
PyTorch Tensor and NumPy ndarray objects can share data storage via array interface. Hence, this issue becomes non-issue if one wraps Tensor object with ndarray.

For instance, when one needs fast indexing on a Tensor object, one can use NumPy fast indexing as illustrated by the following example:
```
>>> import torch, numpy
>>> a=torch.randn(10000)
>>> b=numpy.asarray(a)  # b shares memory with a
>>> import time
>>> start=time.time(); r=[x for x in a]; end=time.time(); print(end-start)
0.01939249038696289
>>> start=time.time(); r=[x for x in b]; end=time.time(); print(end-start)
0.004763603210449219
```

**Comment by kumar-akshay324** [comment]
This is exactly what I am doing. But Pytorch is expected to be handling this better. I had a magnitude lower performance with the tensor on the GPU and the numpy array on the CPU

**Comment by pearu** [comment]
If tensor data is stored on the GPU and you'll need to explicitly iterate over the elements, then it would be more performant to first move tensor to CPU for iterations rather than moving the tensor values from GPU to CPU one by one.

**Comment by vadimkantorov** [comment]
Again hit by this. huge difference between multi-dim scalar indexing into a tensor and indexing into python list :(

**Comment by vadimkantorov** [comment]
Could Python's [`memoryview`](https://docs.python.org/3/library/stdtypes.html#memoryview) be used for indexing into a PyTorch CPU array?

**Comment by vadimkantorov** [comment]
@ngimel maybe a solution could be supporting methods get/set that would directly return/accept Python scalars without intermediate allocation of scalar tensor?

**Comment by ezyang** [comment]
Yes, that would work, and be fairly simple to do (albeit at cost of having to write a lot of code :( )

**Comment by vadimkantorov** [comment]
For get - maybe one can just add argument to item, then it would be `tensor.item(i, j)`

For set, if regular assignment is fast, and maybe accepts tuples / lists without forcing a temporary tensor construction, that would solve it as well.

**Comment by vadimkantorov** [comment]
adding better memoryview support for tensors (with examples) would also be nice!

**Comment by vadimkantorov** [comment]
@ezyang Or maybe even `tensor.item` could be made into an object that has `__call__()` to emulate the old behavior, but also support `__getitem__` to return directly item.

Also, I wonder, if PyTorch could increase Python's integer cache to more values - e.g. to `[-65536; 65536]` instead of default `[-5; 256]`

**Comment by lelouedec** [comment]
I encountered the same problem concerning indexing extremely slow compared to numpy for the same number of computations. 
I used the example of computing the Euclidian distance between all points in a set. Pytorch on CPU is on average around 100 slower for a stupid loop and indexing. 
So decided to try @akamaus strategy on coding it in C++ using directly the data instead of the pytorch C++ tensors with a pointer on the data as : 
`auto  xyz1   = xyz1p.contiguous().data_ptr<float>();`
For the same number of points and high numbers (around 1000), it performed better than both numpy and pytorch by large margins. 
It seems than pytorch when indexing is adding some other computations in the background slowing considerably the process.

**Comment by albanD** [comment]
Hi!

The indexing of Tensor is a lot more expensive because all the intermediary objects are Tensors, Contrary to your plain pointer for loop.
This is expected, you can do autograd, advanced operations, type promotion, etc on these intermediary that you wouldn't be able to do with you plain point for loop.

**Comment by lelouedec** [comment]
> Hi!
> 
> The indexing of Tensor is a lot more expensive because all the intermediary objects are Tensors, Contrary to your plain pointer for loop. This is expected, you can do autograd, advanced operations, type promotion, etc on these intermediary that you wouldn't be able to do with you plain point for loop.

yes of course when comparing these directly to tensor object operations. 
But if we refer to the first posts of the issue with indexing over tensors compare to numpy, the problem is still there. If I do a stupid double for loops using tensors or numpy arrays, the tensor version is systematically slower compared to numpy by a large margin. 
Others have proposed the use of moving into numpy arrays and back to tensor to gain some performances, but doesn't it break the graph and forbid backward/gradient computation?

**Comment by albanD** [comment]
> but doesn't it break the graph and forbid backward/gradient computation?

Well yes. If you want autograd, maybe GPU and all other features that numpy does not have, that has a cost.
The problem is that the indexing ops are usually very cheap. And so all the overhead from the framework related to these features is especially visible.

I am not saying that the advanced indexing in pytorch is as good as it can be though. Just trying to add more context.
